### PR TITLE
variant: visit lambda and value-initialize by index

### PR DESCRIPTION
**Pull request by jeffro256**
Two new features for `tools::variant`:
* Visit by lambda instead of static visitor (possible now because result type is deduced)
* Value-initialize the variant based on a type index

### CODE REVIEW

File: src/common/variant.h [hunk 1]
```diff
@@ -148,16 +161,26 @@ class variant final
...
     }
     template <typename VisitorT>
-    typename VisitorT::result_type visit(VisitorT &&visitor) const
+    decltype(auto) visit(VisitorT &&visitor) const // decltype(auto) since it forwards the return ref type correctly
     {
         return boost::apply_visitor(std::forward<VisitorT>(visitor), m_value);
     }
 
+    /// value initialize the variant based on a type index
+    void value_initialize_to_type_index(const int which)
```
Comment (vtnerd) [inline src/common/variant.h:175]: In what situations would you need to construct based on index rather than type?

Comment (jeffro256) [reply src/common/variant.h:175]: deserialization

Comment (jeffro256) [reply src/common/variant.h:175]: Something similar is already done in the current variant de serialization code:  https://github.com/monero-project/monero/blob/059028a30a8ae9752338a7897329fe8012a310d5/src/serialization/variant.h#L74-L80

**Review by krypt0x — APPROVED** [review]

**Comment by tobtoht** [comment]
The tests introduced here fail on master. Incompatible with #9176?

https://github.com/tobtoht/monero/actions/runs/12470659635/job/34807029624#step:9:4466

**Comment by UkoeHB** [comment]
@jeffro256 can you look into it?

**Pull request merged**